pkg/controller/service/service_controller.go

```diff
@@ -720,47 +617,104 @@ func (s *ServiceController) syncService(key string) error {
 
 	// service holds the latest service info from apiserver
 	service, err := s.serviceLister.Services(namespace).Get(name)
-	switch {
-	case errors.IsNotFound(err):
-		// service absence in store means watcher caught the deletion, ensure LB info is cleaned
-		glog.Infof("Service has been deleted %v. Attempting to cleanup load balancer resources", key)
-		err = s.processServiceDeletion(key)
-	case err != nil:
-		glog.Infof("Unable to retrieve service %v from store: %v", key, err)
-	default:
-		cachedService = s.cache.getOrCreate(key)
-		err = s.processServiceUpdate(cachedService, service, key)
+	if errors.IsNotFound(err) {
+		glog.V(4).Infof("Service %q has been deleted", key)
+		return nil
 	}
 
-	return err
+	if err != nil {
+		glog.Errorf("Unable to retrieve service %v from store: %v", key, err)
+		return err
+	}
+
+	if isDeletionCandidate(service) {
+		err := s.processServiceDeletion(service, key)
+		if err != nil {
+			return err
+		}
+
+		service, err = s.removeFinalizer(service)
+		if err != nil {
+			return err
+		}
+
+		// The service and all external resources, if applicable, have been removed.
+		return nil
+	}
+
+	if needToAddFinalizer(service) {
+		service, err = s.addFinalizer(service)
+		if err != nil {
+			return err
+		}
+
+		// Continue processing the service now that it has been updated with the finalizer.
+	}
+
+	return s.processServiceUpdate(service, key)
 }
 
 // Returns an error if processing the service deletion failed, along with a time.Duration
 // indicating whether processing should be retried; zero means no-retry; otherwise
 // we should retry after that Duration.
-func (s *ServiceController) processServiceDeletion(key string) error {
-	cachedService, ok := s.cache.get(key)
-	if !ok {
-		glog.Errorf("service %s not in cache even though the watcher thought it was. Ignoring the deletion", key)
-		return nil
-	}
-	return s.processLoadBalancerDelete(cachedService, key)
+func (s *ServiceController) processServiceDeletion(service *v1.Service, key string) error {
+	return s.processLoadBalancerDelete(service, key)
 }
 
-func (s *ServiceController) processLoadBalancerDelete(cachedService *cachedService, key string) error {
-	service := cachedService.state
-	// delete load balancer info only if the service type is LoadBalancer
-	if !wantsLoadBalancer(service) {
-		return nil
-	}
+func (s *ServiceController) processLoadBalancerDelete(service *v1.Service, key string) error {
+	glog.Infof("Deleting existing load balancer for service %s that no longer needs a load balancer.", key)
 	s.eventRecorder.Event(service, v1.EventTypeNormal, "DeletingLoadBalancer", "Deleting load balancer")
 	err := s.balancer.EnsureLoadBalancerDeleted(context.TODO(), s.clusterName, service)
```

Comment:
Just realized we may have a problem here. We used to pass in the cached service (previous state), and now we are passing in the updated service (current state). If any cloud provider implementation (e.g. GCE) depends on such behavior, the deletion logic might not work well in case of changing a service from `type=LoadBalancer` to another.

Imagine we change a service from `LoadBalancer` to `ClusterIP`. The old behavior is:
- Cached service object is passed into `EnsureLoadBalancerDeleted()`.
- Cloud provider codes delete LB resources based on the information available on the cached state (from `type`, `externalTrafficPolicy`, `annotations` etc.).
- Clean up the cached service.

While the new behavior is:
- Updated service object is passed into `EnsureLoadBalancerDeleted()`.
- Cloud provider codes may have issue figuring out what resources should be cleaned up as the previous state is no longer available on service.

Reply:
Oops, easy to lose track when I haven't touched the codes for a while.

Seems like the behavior is the same --- we've been passing in the updated service:
https://github.com/kubernetes/kubernetes/blob/3e52ea8081abc13398de6283c31056cd6aecf6b4/pkg/controller/service/service_controller.go#L249-L251